app/services/update_type_service.rb

```diff
@@ -42,19 +42,36 @@ def call(params)
 
   def set_params_and_validate(params)
     # Set patterns includes a data validation before assigning the value to the attribute.
-    # A failure should return a failure.
-    set_patterns(params) if params[:patterns].present?
-    return [false, type.errors] if type.errors.any?
+    # A validation failure should return a service call failure.
+    if params[:patterns].present?
+      validate_enterprise_action(params[:patterns])
+      set_patterns(params[:patterns])
+      return [false, type.errors] if type.errors.any?
+    end
 
     super
   end
 
-  def set_patterns(params)
+  def validate_enterprise_action(patterns)
+    change_from_manual_to_generated = !type.patterns.subject.enabled? && patterns.dig(:subject, :enabled) == true
+
+    if change_from_manual_to_generated && !EnterpriseToken.allows_to?(enterprise_action)
+      type.errors.add(:patterns, :error_enterprise_only, action: enterprise_action.to_s.titleize)
+    end
+  end
+
+  def enterprise_action = :work_package_subject_generation
```

Comment:
:yellow_circle: I find it surprising that the `UpdateTypeService` has a method `enterprise_action` that's specific to subject generation.

Couldn't there be multiple enterprise features on the `Type` that would need to be checked in the `UpdateTypeService`?

I'd suggest one of the following:
* Make `validate_enterprise_action` accept the action as an argument and pass it from `set_params_and_validate` (inside the if block it's clear about which feature we must be talking)
* Rename `enterprise_action` to something like `subject_generation_enterprise_action`

Reply:
Well, usually I'd say this is only important once the service actually DOES check for multiple enterprise actions. But I also have no big concern about using the proposed name.